### PR TITLE
L2 Onboarding

### PR DESCRIPTION
**Pull request by AndyTCF**

### CODE REVIEW

File: src/components/L2Onboarding/StatBox.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	font-size: 24px;
+	line-height: 24px;
+	text-align: center;
+	letter-spacing: 0.2px;
+	color: #ffffff;
+	width: 50px;
+	margin: 8px;
+`;
+
+const Flex = styled.div`
```
Comment (clementbalestrat) [inline src/components/L2Onboarding/StatBox.tsx:74]: There is already a file with basic styles, have a look at `styles/common.js`

File: src/pages/L2Onboarding/Burn.tsx [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+import { getCurrentGasPrice } from 'ducks/network';
+import { bytesFormatter, secondsToTime } from 'helpers/formatters';
+import snxJSConnector from 'helpers/snxJSConnector';
+import GasIndicator from 'components/L2Onboarding/GasIndicator';
+import ErrorMessage from '../../components/ErrorMessage';
+import { addSeconds, differenceInSeconds } from 'date-fns';
+import errorMapper from 'helpers/errorMapper';
+import { Subtext } from 'components/Typography';
+import { useGetDebtData } from './hooks/useGetDebtData';
+import { useGetGasEstimate } from './hooks/useGetGasEstimate';
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/Burn.tsx:19]: you probably don't need to export that do you? It's just gonna be used in this screen.

File: src/pages/L2Onboarding/BurnIntermediary.tsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	walletBalancesWithRates: any;
+}
+
+const BurnIntermediary: React.FC<BurnIntermediaryProps> = ({
+	totalsUSDDebt,
+	setRedirectToTrade,
+	walletBalancesWithRates,
+}) => {
+	const [usdValueOfSynths, setUSDValueOfSynths] = useState<number>(0);
+	const calculateUSDSumOfSynths = useCallback(() => {
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/BurnIntermediary.tsx:25]: You don't need `useCallback` here. Just put it in`useEffect`

File: src/pages/L2Onboarding/BurnIntermediary.tsx [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	}, [calculateUSDSumOfSynths]);
+	const handleRedirectToTrade = () => {
+		setRedirectToTrade(true);
+	};
+	const buttons = [
+		{
+			icon: <TradeIcon />,
+			title: 'TRADE SYNTHS',
+			subtitle: `Balance: $${usdValueOfSynths} USD`,
+			onClick: () => handleRedirectToTrade(),
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/BurnIntermediary.tsx:48]: Nice!

File: src/pages/L2Onboarding/Deposit.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	currentGasPrice,
+	walletBalances,
+}) => {
+	const { currentWallet } = walletDetails;
+	const [snxBalance, setSNXBalance] = useState<number>(0);
+	const [isFetchingGasLimit, setFetchingGasLimit] = useState(false);
+	const [gasLimit, setGasLimit] = useState(0);
+	const sUSDBytes = bytesFormatter('sUSD');
+	const debtData = useGetDebtData(currentWallet, sUSDBytes);
+	const gasEstimateError = useGetGasEstimate(
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/Deposit.tsx:38]: Not needed here. You will need to run estimate against the `transfer` function.

File: src/pages/L2Onboarding/Deposit.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		debtData.sUSDBalance,
+		debtData.maxBurnAmount,
+		debtData.maxBurnAmountBN,
+		debtData.sUSDBalance,
+		null,
+		null,
+		setFetchingGasLimit,
+		setGasLimit
+	);
+	const getSNXBalance = useCallback(async () => {
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/Deposit.tsx:48]: `useCallback` not needed. Put everything in `useEffect`

File: src/pages/L2Onboarding/L2Onboarding.tsx [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+export const L2Onboarding: React.FC<L2OnboardingProps> = ({ setCurrentPage, walletDetails }) => {
+	const [step, setStep] = useState<number>(0);
+	const [sufficientBalance, setSufficientBalance] = useState<boolean | string>('');
+	const [checkingBalances, setCheckingBalances] = useState<boolean>(true);
+	const [sUSDBalance, setSUSDBalance] = useState<number>(0);
+	const { currentWallet } = walletDetails;
+	const sUSDBytes = bytesFormatter('sUSD');
+	const debtData = useGetDebtData(currentWallet, sUSDBytes);
+
+	const fetchDepotData = useCallback(async () => {
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/L2Onboarding.tsx:35]: Same. Put everything back into `useEffect`

File: src/pages/L2Onboarding/hooks/useGetDebtData.tsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
+import { useState, useEffect } from 'react';
+import { bytesFormatter, bigNumberFormatter } from 'helpers/formatters';
+import snxJSConnector from 'helpers/snxJSConnector';
+
+export const useGetDebtData = (walletAddress: string, sUSDBytes: string) => {
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/hooks/useGetDebtData.tsx:5]: I'm quite sure you can have all the debt data from Redux. Have a look at `DebtStatus`

File: src/pages/L2Onboarding/hooks/useGetGasEstimate.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
+import { useState, useEffect } from 'react';
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/hooks/useGetGasEstimate.tsx:1]: As said earlier, you can probably get rid of that file.

File: src/screens/MintrTabs/Home/Home.js [hunk 10]
```diff
@@ -29,9 +30,19 @@ const actionLabelMapper = {
...
 
-const Home = ({ walletDetails: { networkId } }) => {
+const Home = ({ walletDetails: { networkId }, redirectToTrade, setRedirectToTrade }) => {
 	const { t } = useTranslation();
-	const [currentScenario, setCurrentScenario] = useState(initialScenario);
+	const [currentScenario, setCurrentScenario] = useState(
+		redirectToTrade ? 'trade' : initialScenario
+	);
+
+	useEffect(() => {
```
Comment (clementbalestrat) [inline src/screens/MintrTabs/Home/Home.js:39]: maybe call `setRedirectTrade(false)` directly from the `Trade` screen.

File: src/screens/MintrTabs/LPRewards/LPRewards.js [hunk 11]
```diff
@@ -231,13 +231,13 @@ const StyledDataLarge = styled(DataLarge)`
...
 `;
 
-const DistributionRow = styled.div`
-	display: flex;
-	flex-direction: column;
-	& > :not(:first-child) {
-		margin-top: 10px;
-	}
-`;
+// const DistributionRow = styled.div`
```
Comment (clementbalestrat) [inline src/screens/MintrTabs/LPRewards/LPRewards.js:234]: To delete if not needed

File: src/styles/l2/l2colors.ts [hunk 12]
```diff
@@ -0,0 +1,79 @@
+interface IL2Colors {
```
Comment (clementbalestrat) [inline src/styles/l2/l2colors.ts:1]: Maybe rename that ILayer2Icon. Because then you've got plenty of `l1, l2, l3` which can be confusing.

**Review by clementbalestrat — CHANGES_REQUESTED** [review]
Nice job!

You kinda reinvented the wheel sometimes, so I'm sure you can remove plenty of code to be reusing existing components as indicated in my comments.

File: src/components/L2Onboarding/GasSettings.tsx [hunk 13]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (clementbalestrat) [reply src/components/L2Onboarding/GasSettings.tsx:1]: (I was talking about the file itself)

File: src/pages/L2Onboarding/Deposit.tsx (hunk 5 above)
Comment (AndyTCF) [reply src/pages/L2Onboarding/Deposit.tsx:38]: Need clarification for this, but we can talk about it when the OVM address is ready :)

File: src/pages/L2Onboarding/Success.tsx [hunk 14]
```diff
@@ -36,3 +37,8 @@ const PageContainer = styled.div`
 	justify-content: center;
 	align-items: center;
 `;
+
+const CTAButton = styled(ButtonPrimary)`
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/Success.tsx:41]: Maybe export that style into a `component` folder sitting in `/L2Onboarding` so you can reuse it everywhere.

File: src/pages/L2Onboarding/BurnIntermediary.tsx [hunk 15]
```diff
@@ -35,10 +35,21 @@ const BurnIntermediary: React.FC<BurnIntermediaryProps> = ({
 	}, [walletBalancesWithRates]);
 
 	useEffect(() => {
+		const calculateUSDSumOfSynths = () => {
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/BurnIntermediary.tsx:38]: you can just do:

`if (!walletBalancesWithRates) return`

and then remove:

`if (walletBalancesWithRates)...`

File: src/pages/L2Onboarding/Burn.tsx [hunk 16]
```diff
@@ -7,39 +7,146 @@ import { HeaderIcon } from 'components/L2Onboarding/HeaderIcon';
...
 }
 
-const Burn: React.FC<BurnProps> = ({ onComplete, walletDetails, currentGasPrice }) => {
+const useGetDebtData = (walletAddress: string, sUSDBytes: string) => {
+	const [data, setData] = useState<any>({});
+	const SNXBytes = bytesFormatter('SNX');
+	useEffect(() => {
+		const getDebtData = async () => {
+			try {
+				const results = await Promise.all([
```
Comment (clementbalestrat) [inline src/pages/L2Onboarding/Burn.tsx:35]: I'm pretty sure you can fetch all the required data from Redux?
debt, synths balances etc..

**Review by clementbalestrat — CHANGES_REQUESTED** [review]
Nice!  few minor comments.

I think you don't need to create `useGetDebtData` and fetch everything from Redux.

File: src/pages/L2Onboarding/Burn.tsx [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+	const [transferableAmount, setTransferableAmount] = useState<number>(0);
+	const [waitingPeriod, setWaitingPeriod] = useState(0);
+	const [issuanceDelay, setIssuanceDelay] = useState(0);
+	const { currentWallet, walletType } = walletDetails;
+	const [isFetchingGasLimit, setFetchingGasLimit] = useState(false);
+	const [gasLimit, setGasLimit] = useState(0);
+	const { t } = useTranslation();
+
+	const useGetGasEstimate = (
+		burnAmount,
```
Comment (AndyTCF) [inline src/pages/L2Onboarding/Burn.tsx:47]: @clementbalestrat removed the maxBurn and maxBurnNB

File: src/pages/L2Onboarding/Burn.tsx [hunk 18]
```diff
@@ -0,0 +1,273 @@
...
+		useEffect(() => {
+			const getGasEstimate = async () => {
+				setError(null);
+				try {
+					if (burnAmount === 0) throw new Error('You have no debt to burn');
+					if (waitingPeriod) throw new Error('Waiting period for sUSD is still ongoing');
+					if (issuanceDelay) throw new Error('Waiting period to burn is still ongoing');
+					if (burnAmount > sUSDBalance) throw new Error('input.error.notEnoughToBurn');
+					setFetchingGasLimit(true);
+					const amountToBurn = snxJSConnector.utils.parseEther(burnAmount.toString());
```
Comment (AndyTCF) [inline src/pages/L2Onboarding/Burn.tsx:64]: Can I use burnAmount in this way? Straight from the debtData and converted to string

File: src/screens/MintrActions/Migrate/OneInchCard.tsx [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+	debtStatus: DebtStatus;
+	walletBalances: any;
+}) => {
+	const quoteCurrencyKey = CRYPTO_CURRENCY_TO_KEY.ETH;
+	const baseCurrencyKey = CRYPTO_CURRENCY_TO_KEY.sUSD;
+	const [isFetchingGasLimit, setFetchingGasLimit] = useState(false);
+	const [gasLimit, setGasLimit] = useState(0);
+	const { swap, oneInchContract } = useOneInch(
+		appIsReady,
+		snxJSConnector.snxJS.contractSettings.signer
```
Comment (jwineman) [inline src/screens/MintrActions/Migrate/OneInchCard.tsx:36]: Is this the best way to get the signer?

Comment (clementbalestrat) [reply src/screens/MintrActions/Migrate/OneInchCard.tsx:37]: yep

Comment (AndyTCF) [reply src/screens/MintrActions/Migrate/OneInchCard.tsx:37]: const {
          signer
	} = snxJSConnector;

Comment (AndyTCF) [reply src/screens/MintrActions/Migrate/OneInchCard.tsx:37]: `	const {
		snxJS: { EtherCollateral, EtherCollateralsUSD },
		signer
	} = snxJSConnector;`

File: src/hooks/useOneInch.ts [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+					sUSDTokenAddress,
+					amountBN,
+					swapRates.returnAmount,
+					swapRates.distribution,
+					0,
+				];
+
+				const tx = oneInchContract.functions.swap(...swapParams, {
+					value: amountBN,
+					gasPrice,
```
Comment (AndyTCF) [inline src/hooks/useOneInch.ts:48]: @jwineman pretty sure gasPrice is also a BN in wei and also use the estimated gas limit from the estimate you did 

example in synthetix exchange 

```	const gasEstimate = await contract.estimate.liquidateLoan(loanOwner, loanIDStr, debtToCover);
			const updatedGasEstimate = normalizeGasLimit(Number(gasEstimate));
			setLocalGasLimit(updatedGasEstimate);

			const tx = await contract.liquidateLoan(loanOwner, loanIDStr, debtToCover, {
				gasPrice: gasInfo.gasPrice * GWEI_UNIT,
				gasLimit: updatedGasEstimate,
			});```

**Pull request merged**